### PR TITLE
[403] Revoke user access to organisations feature

### PR DESCRIPTION
**Pull request by stellenberger**
### Context

- [This trello ticket](https://trello.com/c/XnSq0qEL/403-revoke-user-access-to-organisations-feature)

### Changes proposed in this pull request

- Add functionality to remove providers from users

### Guidance to review

- Log onto support console, navigate to users, select a user, and click the big red button to remove the provider from that user

### Checklist

- [x] Make sure all information from the Trello card is in here
- [x] Attach to Trello card
- [x] Rebased master
- [x] Cleaned commit history
- [x] Tested by running locally

### CODE REVIEW

**Comment by stellenberger** [comment]
[Revoke user access to organisations feature](https://trello.com/c/XnSq0qEL/403-revoke-user-access-to-organisations-feature)

File: app/controllers/support/user_permissions_controller.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
+module Support
+  class UserPermissionsController < SupportController
+    def destroy
+      if user.remove_access_to(provider)
+        redirect_to origin_page, flash: { success: "#{user_full_name} removed from #{provider.provider_name}" }
```
Comment (darokel) [inline app/controllers/support/user_permissions_controller.rb:5]: We've got some generic flash messages for CRUD actions, see the `app/controllers/support/locations_controller.rb` for eg. Would be good to maybe keep this consistent

File: app/controllers/support/user_permissions_controller.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+    def provider
+      @provider ||= user_permission.provider
+    end
+
+    def origin_page
+      request.referer
+    end
+
+    def user_full_name
```
Comment (darokel) [inline app/controllers/support/user_permissions_controller.rb:29]: This sort of presentational stuff shouldn't live in the controller. I think we can avoid it all together if we stick with the generic flash messages

File: app/views/support/providers/_provider_user_list.html.erb [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            <span class="govuk-!-display-block govuk-!-margin-bottom-1"><%= user.first_name %></span>
+        </td>
+
+        <td class="govuk-table__cell">
+          <span class="govuk-!-display-block govuk-!-margin-bottom-1"><%= user.last_name %></span>
+        </td>
+
+        <td class="govuk-table__cell">
+          <span class="govuk-!-display-block govuk-!-margin-bottom-1">
+            <%= govuk_link_to user.email, support_user_path(user) %>
```
Comment (darokel) [inline app/views/support/providers/_provider_user_list.html.erb:24]:
```suggestion
<%= govuk_link_to(user.email, support_user_path(user)) %>
```

File: app/views/support/users/_user_providers.html.erb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+      <th scope="col" class="govuk-table__header govuk-!-width-one-half">Action</th>
+    </tr>
+  </thead>
+
+  <tbody class="govuk-table__body">
+    <% @providers.each do |provider| %>
+      <tr class="govuk-table__row qa-provider_row">
+        <td class="govuk-table__cell">
+          <span class="govuk-!-display-block govuk-!-margin-bottom-1">
+            <%= govuk_link_to provider.provider_name, support_provider_path(provider) %>
```
Comment (darokel) [inline app/views/support/users/_user_providers.html.erb:15]:
```suggestion
<%= govuk_link_to(provider.provider_name, support_provider_path(provider)) %>
```

File: spec/features/support/providers/deleting_user_from_provider_spec.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    expect(provider_users_index_page).to have_content("#{@user.first_name} #{@user.last_name} removed from #{@provider.provider_name}")
+  end
+
+  def then_i_am_taken_to_the_provider_users_index_page
+    expect(provider_users_index_page).to be_displayed
+  end
+
+  def and_the_user_provider_relationship_is_destroyed
+    @provider.reload
+    expect(@provider.users.length).to eq 0
```
Comment (darokel) [inline spec/features/support/providers/deleting_user_from_provider_spec.rb:49]: Would `expect(@provider.users).to be_empty` work here? Same for the `provider_users_index_page.users`

**Review by darokel — COMMENTED** [review]
Looking good overall but left some comments

File: app/controllers/support/user_permissions_controller.rb (hunk 2 above)
Comment (stellenberger) [reply app/controllers/support/user_permissions_controller.rb:29]: Agreed, removed as using the generic flash messages now 👍

File: spec/features/support/providers/deleting_user_from_provider_spec.rb (hunk 5 above)
Comment (stellenberger) [reply spec/features/support/providers/deleting_user_from_provider_spec.rb:49]: Good spot, changed both

**Review by darokel — APPROVED** [review]

**Review by mbocc — APPROVED** [review]
New feature works. If I click on remove provider, it not only removes the provider from the list below the user, but also removes the user from the provider page. 

All works, and looks good - thanks for the work!

**Pull request merged**